drv/sidecar-mainboard-controller/src/tofino2.rs

```diff
@@ -8,29 +8,246 @@ use drv_fpga_api::{FpgaError, FpgaUserDesign, WriteOp};
+            Reg::TOFINO_SEQ_ERROR_STATE::STATE,
+        ))
+        .ok_or(FpgaError::InvalidValue)?;
+
+        let error_step = TofinoSeqStep::from_u8(value(
+            Addr::TOFINO_SEQ_ERROR_STEP,
+            Reg::TOFINO_SEQ_ERROR_STEP::STEP,
+        ))
+        .ok_or(FpgaError::InvalidValue)?;
+
+        Ok(TofinoSeqStatus {
+            state,
+            step,
+            abort: match error {
+                TofinoSeqError::None => None,
+                _ => Some(TofinoSeqAbort {
+                    state: error_state,
+                    step: error_step,
+                    error,
+                }),
+            },
+        })
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq)]
+#[repr(C)]
+pub struct TofinoSeqAbort {
+    pub state: TofinoSeqState,
+    pub step: TofinoSeqStep,
+    pub error: TofinoSeqError,
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum PowerRails {
+    Vdd18,
+    VddCore,
+    VddPcie,
+    Vddt,
+    Vdda15,
+    Vdda18,
+}
+
+#[derive(
+    Copy, Clone, Debug, Default, Eq, PartialEq, FromPrimitive, AsBytes,
+)]
+#[repr(u8)]
+pub enum PowerRailState {
+    #[default]
+    Disabled = 0,
+    RampingUp = 1,
+    GoodTimeout = 2,
+    Aborted = 3,
+    Enabled = 4,
+}
+
+impl TryFrom<u8> for PowerRailState {
+    type Error = FpgaError;
+
+    fn try_from(v: u8) -> Result<Self, Self::Error> {
+        match v {
+            0 => Ok(PowerRailState::Disabled),
+            1 => Ok(PowerRailState::RampingUp),
+            2 => Ok(PowerRailState::GoodTimeout),
+            3 => Ok(PowerRailState::Aborted),
+            4 => Ok(PowerRailState::Enabled),
+            _ => Err(FpgaError::InvalidValue),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Default, Eq, PartialEq)]
+#[repr(C)]
+pub struct PowerRailPinState {
+    pub enable: bool,
+    pub good: bool,
+    pub fault: bool,
+    pub vrhot: bool,
+}
+
+impl From<u8> for PowerRailPinState {
+    fn from(v: u8) -> Self {
+        use Reg::TOFINO_POWER_VDD18_STATE::*;
+
+        Self {
+            enable: v & ENABLE != 0,
+            good: v & GOOD != 0,
+            fault: v & FAULT != 0,
+            vrhot: v & VRHOT != 0,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[repr(C)]
+pub struct PowerRail {
+    pub id: PowerRails,
+    pub state: PowerRailState,
+    pub pin_state: PowerRailPinState,
+}
+
+impl TryFrom<(PowerRails, u8)> for PowerRail {
+    type Error = FpgaError;
+
+    fn try_from(data: (PowerRails, u8)) -> Result<Self, Self::Error> {
+        let (id, v) = data;
+        Ok(PowerRail {
+            id,
+            state: PowerRailState::try_from(v >> 4)?,
+            pin_state: PowerRailPinState::from(v),
+        })
+    }
+}
+
+impl PowerRail {
+    /// Convert power rail state register data to an array of `PowerRail`
+    /// structs.
+    pub fn from_raw(data: [u8; 6]) -> Result<[PowerRail; 6], FpgaError> {
+        #[inline]
+        const fn offset(addr: Addr) -> usize {
```

Comment:
Nit: you could capture `data` here, using the same pattern as in `TofinoSeqStatus`:
```rust
let mut value = |addr: Addr| -> u8 {
    data[(addr as usize) - (Addr::TOFINO_POWER_VDD18_STATE as usize)]
};
```